### ISSUE TITLE
Add microlens-pro with prisms and isomorphisms

### ISSUE THREAD

**Issue by neongreen** [stevenfontanella/microlens#105]
Status update: I started working on it several days ago.

**Comment by bradrn** [comment]
Thank you! I've actually been thinking of opening this as an issue for a while, so it's nice to see you've already thought of it.

**Comment by bradrn** [comment]
Actually, I've just seen that this is from 3 months ago (I didn't see that in my last comment). Do you have any idea as to when this will be merged?

**Comment by neongreen** [comment]
No idea, sorry.

**Comment by neongreen** [comment]
I'm unhappy with how it's turning out to be (there are tradeoffs like "do we reimplement a bunch of classes or do we cripple many of the basic usecases people need prisms for") and so it's hard to bring myself to finish it 🙂

**Comment by bradrn** [comment]
@neongreen Can you elaborate on those tradeoffs? Perhaps if they get discussed here we might be able to find a solution which everyone's happy with. I know I'd be prepared to accept tradeoffs if they don't affect most usecases too much yet still allow me to use Profunctor-based lenses/isos/prisms.

**Comment by neongreen** [comment]
@bradrn Terribly sorry, I was quitting my job and lots of things slipped through my fingers in the past month.

Here's an elaboration on the tradeoffs:

* Most prisms and isos live in typeclasses (`_Cons`, `_head`, `_Empty`, `_Wrapped`, `strict`, `swapped`). Do we implement those typeclasses, or leave users with prisms types but no actual prisms?

* `_Left` and friends are implemented as traversals in microlens, microlens-ghc, microlens-platform, but they are actually prisms. Do we implement them in microlens-pro? If yes, how do we solve name clashes? It seems weird for microlens-pro to clash with microlens. Perhaps microlens-pro should reexport most of microlens but with `hiding`?

(I guess that at this point I would be fine with just providing definitions and essentials like `prism`, but not providing any actual prisms.)

**Comment by bradrn** [comment]
@neongreen I don't know much about this area of `lens`, but here's what I think:

> * Most prisms and isos live in typeclasses (`_Cons`, `_head`, `_Empty`, `_Wrapped`, `strict`, `swapped`). Do we implement those typeclasses, or leave users with prisms types but no actual prisms?

I thought most of the code was just copied from `lens`. Is there any reason _not_ to include those typeclasses?

> * `_Left` and friends are implemented as traversals in microlens, microlens-ghc, microlens-platform, but they are actually prisms. Do we implement them in microlens-pro? If yes, how do we solve name clashes? It seems weird for microlens-pro to clash with microlens. Perhaps microlens-pro should reexport most of microlens but with `hiding`?

I'm thinking that you could adopt a naming convention, something like `_Left` is the traversal, `p_Left` or `_Leftp` is the prism (I like the first better). Or if you really want to preserve compatibility with `lens`, possibly a better alternative is renaming the traversal to `t_Left` and keeping the prism as just `_Left`.

**Comment by neongreen** [comment]
> Is there any reason not to include those typeclasses?

The primary reason is that either we can't provide any instances, or we have to depend to `bytestring`, `text`, `vector`, `unordered-containers`.

> Or if you really want to preserve compatibility with lens, possibly a better alternative is renaming the traversal to `t_Left` and keeping the prism as just `_Left`.

This still breaks compatibility with lens.

One of goals of `microlens` is that at any time you should be able to swap it with `lens` and everything would still work. We slightly break this with `SimpleGetter`, which is a necessary evil, but it will go away when GHC 8.10 is out (https://github.com/monadfix/microlens/issues/102).

**Comment by neongreen** [comment]
Oops, closed accidentally.

**Comment by bradrn** [comment]
> > Is there any reason not to include those typeclasses?
> 
> The primary reason is that either we can't provide any instances, or we have to depend to `bytestring`, `text`, `vector`, `unordered-containers`.

It looks like there's still a number of useful isomorphisms/prisms which don't require these dependencies, like `_Left`, `_Right`, `_Wrapped` etc. For the ones which do require a dependency, I would think a nice solution would be to define the typeclasses in `microlens-pro`, with a few simple instances e.g. lists, and then define the instances in `microlens-ghc`. Or if you don't want a `profunctor` dependency for `microlens-ghc` make a new package `microlens-pro-ghc` (although I personally think that's unnecessary and excessively modular).

> > Or if you really want to preserve compatibility with lens, possibly a better alternative is renaming the traversal to `t_Left` and keeping the prism as just `_Left`.
> 
> This still breaks compatibility with lens.
> 
> One of goals of `microlens` is that at any time you should be able to swap it with `lens` and everything would still work. We slightly break this with `SimpleGetter`, which is a necessary evil, but it will go away when GHC 8.10 is out (#102).

Yes, that seems like a good point. Thinking about it some more, I think it is reasonable to require users to import `microlens-pro` in order to use these combinators.

**Comment by neongreen** [comment]
I like the idea of microlens-platform depending on microlens-pro, but our users might not. (And yeah, having microlens-pro-ghc, microlens-pro-platform, etc would be excessive.)

I don't think it's reasonable to remove `_Left`, `_Right` and so on from microlens and require users to import microlens-pro. For me, the most important role of microlens is promoting the usage of lenses in library code, and anything that cripples microlens's ability to do that is a no-go.

**Comment by bradrn** [comment]
@neongreen I think that summarises quite well the essential tradeoffs involved in `microlens`. _Either_ we remain compatible with `lens`, _or_ we make `microlens` easy to use. Going forward I think that someone has to decide which part is more important.

> I like the idea of microlens-platform depending on microlens-pro

I never mentioned microlens-platform. Anyway, I thought that microlens-platform was meant to depend on everything, so I assumed that it would depend on microlens-pro as well.

**Comment by neongreen** [comment]
> I never mentioned microlens-platform.

You mentioned microlens-ghc, and everything that applies to microlens-ghc also applies to microlens-platform. The "platform" in microlens-platform stands for "Haskell platform", not "all microlens packages".

In hindsight I can see how that would be confusing 🙂 Perhaps this could be changed starting from 1.0.

**Comment by bradrn** [comment]
> The "platform" in microlens-platform stands for "Haskell platform", not "all microlens packages".

I didn't know this! I thought it was just all the microlens packages together. It's probably worth pointing this out in the README or somewhere else prominent.

----

Anyway, it looks like we've gotten a bit off-topic 🙂 I've been thinking a bit about the design of `microlens-pro`, and I think I've stabilised on my own 'ideal' `microlens-pro` implementation:

> Add a new package `microlens-pro` for prisms/isos/anything else requiring the `profunctor` package. This package will contain everything related to prisms/isos which depends only on `base` and `profunctor`. So it will contain `Prism`, `Iso`, `_Left`, `_Right`, `strict` etc., as well as typeclasses `Strict`, `Reversing` etc. with `base` instances like `[]`. All other instances will be placed into `microlens-ghc` or `microlens-platform` as appropriate. The existing traversals/lenses in `microlens` will be renamed to `t_Left`, `t_Right`, `l_strict`, `l_lazy` etc. (using a convention traversal→`t` and lens→`l`; I'd be fine with other conventions as well).

The only problem I can see with this proposal is renaming the prism-/iso-substitutes in `microlens`. Personally, I don't see this as a particularly major problem: the assumption is that if the user does use these, then they are using them with the understanding that they are not included in `lens`. Probably a disclaimer along these lines in the relevant documentation would assist.

Alternately, if you really _don't_ want to break compatibility with `lens`, I did think of another solution which would work without renaming anything. You could define _both_ sets of prisms/isos without renaming anything, then in `microlens-pro` define a new module `Lens.Micro.ProCompat` which simply re-exports everything from `Lens.Micro` but hiding `_Left`, `strict` etc. The intention is that the user could do this:

```haskell
import Lens.Micro.ProCompat
import Lens.Micro.Pro
```

Instead of this:

```haskell
import Lens.Micro
import Lens.Micro.Pro
```

I don't particularly like this solution - I think it's a bit clunky - but I do think it's a nicer option for retaining `lens`-compatibility than anything we've suggested so far.

**Comment by neongreen** [comment]
By the way, here's the current dependency list of `profunctors`:

```
+Cabal 2.4.1.0
+StateVar 1.1.1.1
+array 0.5.3.0
+base-orphans 0.8.1
+bifunctors 5.5.3
+binary 0.8.6.0
+bytestring 0.10.8.2
+cabal-doctest 1.0.6
+comonad 5.0.4
+containers 0.6.0.1
+contravariant 1.5
+deepseq 1.4.4.0
+directory 1.3.3.0
+distributive 0.6
+filepath 1.4.2.1
+ghc-boot-th 8.6.4
+mtl 2.2.2
+parsec 3.1.13.0
+pretty 1.1.3.6
+process 1.6.5.0
+profunctors 5.3
+semigroups 0.18.5
+stm 2.5.0.0
+tagged 0.8.6
+template-haskell 2.14.0.0
+text 1.2.3.1
+th-abstraction 0.2.11.0
+time 1.8.0.2
+transformers 0.5.6.2
+transformers-compat 0.6.4
+unix 2.7.2.2
```

It means that `microlens-pro` would depend on everything from Haskell Platform anyway, really, so we can put all those instances into `microlens-pro`.

However, the dependency situation might change once things like https://github.com/ekmett/comonad/issues/47 are fixed – which will happen who-knows-when.

Given all this, I'm inclined to say "if you use microlens-pro, that's the only package you need; don't import microlens or microlens-platform". Then we can do a bunch of `hiding` internally and our users don't have to worry about it.

It will also confuse the heck out of everyone because people wouldn't know if "pro" means "profunctors" or "professional". He-he.

**Comment by bradrn** [comment]
That's interesting - I knew `profunctors` was a large library, but I didn't know it was _that_ large. In that case, if `profunctors` already depends on most of the Haskell Platform, then is there anything stopping us from simply putting prisms/isos into `microlens-platform`?

**Comment by neongreen** [comment]
> then is there anything stopping us from simply putting prisms/isos into microlens-platform?

The cost of profunctors+comonad+bifunctors+distributive is non-trivial. Here are build times for some configurations (on a recent MacBook Pro with GHC 8.6.4):

|library|completely from scratch|with common deps preinstalled\*|with common deps+Cabal preinstalled|
|-|-|-|-|
|microlens-platform|2m30s|20s|20s|
|microlens-platform with profunctors|7m20s|5m20s|55s|
|lens|9m00s|7m30s|2m45s|

\* common deps = bytestring, vector, text, containers, unordered-containers, mtl

Even with `Cabal` preinstalled, adding `profunctors` to `microlens-platform` almost triples the compilation time. I'd rather not do that.

**Comment by bradrn** [comment]
> I'd rather not do that.

Yes, I agree; it was a stupid idea in the first place.

----

Anyway, since `profunctors` does come with most of the platform, it would make sense then to define all the instances in `microlens-pro` anyway. In general, I'm inclined to agree with your comment from earlier:

> "If you use microlens-pro, that's the only package you need; don't import microlens or microlens-platform."

To clarify: are you suggesting that `microlens-pro` should re-export `microlens-platform` here?

**Comment by neongreen** [comment]
Yes, reexport most of it (except for things like _Left, which will have to be hidden and redefined)

**Comment by bradrn** [comment]
So to confirm: the final plan is to re-export all of `microlens-platform` from `microlens-pro`, but hiding name clashes?

To be honest, if that's the final plan, I don't really see the point of even making `microlens-pro`; in this case the dependencies and aggregate compile time will be so similar to `lens` as to make no difference.

**Comment by neongreen** [comment]
> in this case the dependencies and aggregate compile time will be so similar to lens as to make no difference.

If you happen to have `Cabal` or if you disable it with `-f-test-doctests`, it's still three times faster than building `lens`. Plus you get a smaller library that is easier for others to study – if I have a project that uses `microlens-pro`, I can just tell people "hey, read the docs". I can not do that with `lens`.

I can not really see using `microlens-pro` for my own libraries, but I know that there are people who would use it even though they transitively depend on most of the `lens` deps already.

**Comment by bradrn** [comment]
> If you happen to have `Cabal` or if you disable it with `-f-test-doctests`, it's still three times faster than building `lens`.

Won't `lens` compilation be faster as well if you do this?

>Plus you get a smaller library that is easier for others to study – if I have a project that uses `microlens-pro`, I can just tell people "hey, read the docs". I can not do that with `lens`.

Why not? Admittedly, the documentation for `microlens` is several times better than that of `lens`, but considering the average state of Haskell documentation, I would argue that this doesn't constitute a barrier for the use of `lens`.

**Comment by neongreen** [comment]
> Won't lens compilation be faster as well if you do this?

I took that into account (see the table above) – with `Cabal` preinstalled it is 55s for `microlens-platform + profunctors`, vs. 2m45s for `lens`.

> I would argue that this doesn't constitute a barrier for the use of lens.

If you already want to use `lens`, then sure. If you have to push for it (in a `lens`-averse workspace, for instance), it helps to have as many arguments for your side as possible.

**Comment by bradrn** [comment]
> I took that into account (see the table above) – with `Cabal` preinstalled it is 55s for `microlens-platform + profunctors`, vs. 2m45s for `lens`.

That's an amazing difference - I would definitely use `microlens-pro` if this remains true! Do you have any idea why this happens?

> If you already want to use `lens`, then sure. If you have to push for it (in a `lens`-averse workspace, for instance), it helps to have as many arguments for your side as possible.

True 🙂

**Comment by neongreen** [comment]
> Why this happens

1. `lens` is big.

2. `lens` has some extra dependencies that are not in the `profunctors` dependency graph:

   ```
   adjunctions 4.4
   call-stack 0.1.0
   exceptions 0.10.1
   free 5.1
   invariant 0.5.2
   kan-extensions 5.2
   parallel 3.2.2.0
   reflection 2.1.4
   semigroupoids 5.3.2
   transformers-base 0.4.5.2
   void 0.7.2
   ```

**Comment by bradrn** [comment]
That makes sense.

---

To get back on topic, I'm wondering:

1. Would you say that at this point the design of `microlens-pro` is fairly settled?
2. If the answer to (1) is 'yes', when do you expect `microlens-pro` to be released (on either GitHub or Hackage)?

**Comment by neongreen** [comment]
1. I think the design is final, yes (unless somebody else chimes in).

2. I'd rather not give an estimate after missing the previous one by
several months. A release won't happen before GHC 8.8 is done, but
unfortunately I don't have an upper bound. One thing I can do is merge the
microlens-pro PR sooner, and create issues for non-essential features.

**Comment by bradrn** [comment]
Thanks @neongreen!

**Comment by lrworth** [comment]
@neongreen I keep looking for this package. If you point me at a WIP branch I can help develop a PR.

**Comment by stevenfontanella** [comment]
@lrworth I see that you already worked around this, but feel free to send a PR in the future if interested.

**Comment by crumbtoo** [comment]
Hi! Could I pick up on this if nobody else is doing it?

**Comment by stevenfontanella** [comment]
Absolutely! Please feel free to use the draft PR #107 as a reference if you find it useful.

**Comment by stevenfontanella** [comment]
Also, please share a draft PR when you start work on this. Emily (neongreen) mentioned some tradeoffs in this issue, so I want to know which choices we're making. Feel free to shoot me an email me as well to discuss this: steven.fontanella@gmail.com.

**Comment by crumbtoo** [comment]
Apologies if this is a second notification! Thought it might be worth mentioning the draft PR #174 here.

**Comment by stevenfontanella** [comment]
Published [microlens-pro 0.1](https://hackage.haskell.org/package/microlens-pro-0.1.0). Thanks again @crumbtoo and @neongreen!

**Comment by neongreen** [comment]
Congrats everyone!

**Comment by neongreen** [comment]
(Reddit announcement coming?)

**Comment by stevenfontanella** [comment]
Yes!

**Comment by stevenfontanella** [comment]
Let me publish a new version with the changes in https://github.com/stevenfontanella/microlens/pull/174#issuecomment-1925591528, then I'll announce it after.

**Comment by stevenfontanella** [comment]
Since `profunctors` adds a lot of dependencies, I was curious how `microlens-pro` compares to the full `lens`. I got these transitive dependency lists for [microlens-pro](https://gist.github.com/stevenfontanella/7d26d877b48a8cc8de16e41af2cfae8e) and [lens](https://gist.github.com/stevenfontanella/206a44b914e522e824ce7c4b459588a0), with a diff of [16 packages](https://gist.github.com/stevenfontanella/46a8e5975dbca0c823c3434fcb1dde1e).

I got it with `cabal v2-freeze`. Since it includes the whole `cabal.project`, I only included `packages: .` and removed all `test-suite` stanzas from `lens.cabal` before running it to avoid including dependencies like HUnit.
